File: extensions/youtube/src/main/java/app/morphe/extension/youtube/patches/CustomPlayerOverlayOpacityPatch.java
package app.morphe.extension.youtube.patches;

import static app.morphe.extension.shared.StringRef.str;

import android.widget.ImageView;

import app.morphe.extension.shared.Utils;
import app.morphe.extension.youtube.settings.Settings;

@SuppressWarnings("unused")
public class CustomPlayerOverlayOpacityPatch {

    private static final int PLAYER_OVERLAY_OPACITY_LEVEL;

    static {
        int opacity = Settings.PLAYER_OVERLAY_OPACITY.get();

        if (opacity < 0 || opacity > 100) {
            Utils.showToastLong(str("morphe_player_overlay_opacity_invalid_toast"));
            opacity = Settings.PLAYER_OVERLAY_OPACITY.resetToDefault();
        }

        PLAYER_OVERLAY_OPACITY_LEVEL = (opacity * 255) / 100;
    }

    /**
     * Injection point.
     */
    public static void changeOpacity(ImageView imageView) {
        imageView.setImageAlpha(PLAYER_OVERLAY_OPACITY_LEVEL);
    }
}

File: patches/src/main/kotlin/app/morphe/patches/youtube/layout/player/overlay/CustomPlayerOverlayOpacityPatch.kt
package app.morphe.patches.youtube.layout.player.overlay

import app.morphe.patcher.extensions.InstructionExtensions.addInstruction
import app.morphe.patcher.extensions.InstructionExtensions.getInstruction
import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patches.all.misc.resources.resourceMappingPatch
import app.morphe.patches.shared.misc.settings.preference.InputType
import app.morphe.patches.shared.misc.settings.preference.TextPreference
import app.morphe.patches.youtube.layout.buttons.overlay.addPlayerOverlayPreferences
import app.morphe.patches.youtube.layout.buttons.overlay.playerOverlayButtonsSettingsPatch
import app.morphe.patches.youtube.misc.settings.settingsPatch
import app.morphe.patches.youtube.shared.Constants.COMPATIBILITY_YOUTUBE
import com.android.tools.smali.dexlib2.iface.instruction.OneRegisterInstruction

private const val EXTENSION_CLASS =
    "Lapp/morphe/extension/youtube/patches/CustomPlayerOverlayOpacityPatch;"

@Suppress("unused")
val customPlayerOverlayOpacityPatch = bytecodePatch(
    name = "Custom player overlay opacity",
    description = "Adds an option to change the opacity of the video player background when player controls are visible.",
) {
    dependsOn(settingsPatch,
        resourceMappingPatch,
        playerOverlayButtonsSettingsPatch
    )

    compatibleWith(COMPATIBILITY_YOUTUBE)

    execute {
        addPlayerOverlayPreferences(
            TextPreference("morphe_player_overlay_opacity", inputType = InputType.NUMBER)
        )

        CreatePlayerOverviewFingerprint.let {
            it.method.apply {
                val viewRegisterIndex = it.instructionMatches.last().index
                val viewRegister = getInstruction<OneRegisterInstruction>(viewRegisterIndex).registerA

                addInstruction(
                    viewRegisterIndex + 1,
                    "invoke-static { v$viewRegister }, " +
                            "$EXTENSION_CLASS->changeOpacity(Landroid/widget/ImageView;)V",
                )
            }
        }
    }
}
